src/linpred.jl

```diff
@@ -35,35 +35,49 @@ end
 """
     DensePredQR
 
-A `LinPred` type with a dense, unpivoted QR decomposition of `X`
+A `LinPred` type with a dense QR decomposition (both pivoted and unpivoted) of `X`
 
 # Members
 
 - `X`: Model matrix of size `n` × `p` with `n ≥ p`.  Should be full column rank.
 - `beta0`: base coefficient vector of length `p`
 - `delbeta`: increment to coefficient vector, also of length `p`
 - `scratchbeta`: scratch vector of length `p`, used in `linpred!` method
-- `qr`: a `QRCompactWY` object created from `X`, with optional row weights.
+- `qr`: either a `QRCompactWY` or `QRPrivoted` object created from `X`, with optional row weights.
 """
-mutable struct DensePredQR{T<:BlasReal} <: DensePred
+mutable struct DensePredQR{T<:BlasReal,Q} <: DensePred
     X::Matrix{T}                  # model matrix
     beta0::Vector{T}              # base coefficient vector
     delbeta::Vector{T}            # coefficient increment
     scratchbeta::Vector{T}
-    qr::QRCompactWY{T}
-    function DensePredQR{T}(X::Matrix{T}, beta0::Vector{T}) where T
-        n, p = size(X)
-        length(beta0) == p || throw(DimensionMismatch("length(β0) ≠ size(X,2)"))
-        new{T}(X, beta0, zeros(T,p), zeros(T,p), qr(X))
-    end
-    function DensePredQR{T}(X::Matrix{T}) where T
-        n, p = size(X)
-        new{T}(X, zeros(T, p), zeros(T,p), zeros(T,p), qr(X))
-    end
+    qr::Q
+end
+function DensePredQR(X::AbstractMatrix, beta0::AbstractVector, pivot::Bool=false)
+    n, p = size(X)
+    length(beta0) == p || throw(DimensionMismatch("length(β0) ≠ size(X,2)"))
+    T = eltype(X)
+    F = pivot ? pivoted_qr(X) : qr(X)
+    DensePredQR(Matrix{T}(X),
+        Vector{T}(beta0),
+        zeros(T, p),
+        zeros(T, p),
+        F)
 end
-DensePredQR(X::Matrix, beta0::Vector) = DensePredQR{eltype(X)}(X, beta0)
-DensePredQR(X::Matrix{T}) where T = DensePredQR{T}(X, zeros(T, size(X,2)))
-convert(::Type{DensePredQR{T}}, X::Matrix{T}) where {T} = DensePredQR{T}(X, zeros(T, size(X, 2)))
+function DensePredQR(X::AbstractMatrix, pivot::Bool=false)
+    n, p = size(X)
+    T = eltype(X)
+    F = pivot ? pivoted_qr(X) : qr(X)
+    DensePredQR(Matrix{T}(X),
+        zeros(T, p),
+        zeros(T, p),
+        zeros(T, p),
+        F)
+end
+
+#DensePredQR(X::Matrix, beta0::Vector, pivot::Bool=false) = DensePredQR{eltype(X)}(X, beta0, pivot)
```

Comment:
why do we have here these commented lines?

Reply:
Removed the code.